utils/litemap/src/map.rs

```diff
@@ -201,33 +201,52 @@ impl<K: Ord, V> LiteMap<K, V> {
     /// Extend the map's underlying vector with the items in the iterator as long as
     /// the appended keys remain in sorted order with the vector's existing keys.
     ///
-    /// If successful, return the map containing the appended items.
-    /// Otherwise return the map and the rest of the unsorted items.
-    fn try_extend_from_sorted<I, E>(mut self, iter: I) -> Result<Self, (Self, E)>
+    /// If all items were appended successfully, return `None`.
+    /// Otherwise, return an `Some` iterator over the rest of the unsorted items.
+    ///```rust
+    /// use litemap::LiteMap;
+    ///
+    /// let mut map = LiteMap::new();
+    /// let iter = [(2, "two"), (1, "one"), (3, "three")].iter().cloned();
+    /// let option = map.try_extend_from_sorted(iter);
+    ///
+    /// assert_eq!(map.get(&2), Some(&"two"));
+    /// assert_eq!(map.get(&1), None);
+    /// assert_eq!(option.unwrap().next(), Some((1, "one")));
+    ///```
+    pub fn try_extend_from_sorted<I>(&mut self, iter: I) -> Option<impl Iterator<Item = (K, V)>>
```

Comment:
I like this a lot better. :)

How about making the return value be

```rust
Result<(), impl Iterator<Item = (K, V)>>
```

Because then you trigger the Rust warnings about ignoring the return value.  If you expect the call to succeed, then you can `.expect()` it.

Reply:
Hmm, perhaps we could/should change `try_append()` to be `Result<(), (K, V)>` instead of `Option<(K, V)>` as well? 

@Manishearth do you have an opinion here? I'm following your style of using `Option` for `try_append()`.

Reply:
You can achieve it with `#[must_use]` - no need for `Result` just to get the warning.

Reply:
That would be fine with me then.  Either return a Result or tag the method with `#[must_use]`.

I still think I like Result a bit better though, because the expectation is that if you call this method, you expect your input to be sorted, and if it's not sorted, that's an error, and you might want to use `?` to unwrap the stack.

Reply:
I've added the `#[must_use]` attribute.

Reply:
This method feels a little awkward, should it just be private? I'm okay exposing this but really I'm not sure if anyone would actually _need_ this. Doesn't really hurt, though.

Reply:
I think the method is useful.  I frequently deal with data that is pre-sorted, and I need to transfer it into a litemap.

Reply:
Even if it is useful in some situations, I agree that it is a bit awkward. Plus, a simpler solutions exists, that also provides more granular optimal performance for `collect()`, as you pointed out in your other comment, @sffc  

For now, I've updated the PR to remove the two `extend_from` functions, since they're no longer needed. 

If we decide that they are helpful enough to be in the public API later, all the code is here and/or simple enough to re-write quickly in a separate PR.